crates/sui-types/src/messages.rs

```diff
@@ -1323,40 +1323,37 @@ impl From<VerifiedObjectInfoResponse> for ObjectInfoResponse {
     }
 }
 
-#[derive(Debug, PartialEq, Eq, Hash, Clone, Serialize, Deserialize)]
+#[derive(Clone, Debug, Serialize, Deserialize)]
 pub struct TransactionInfoRequest {
     pub transaction_digest: TransactionDigest,
 }
 
-impl From<TransactionDigest> for TransactionInfoRequest {
-    fn from(transaction_digest: TransactionDigest) -> Self {
-        TransactionInfoRequest { transaction_digest }
-    }
-}
-
 #[derive(Clone, Debug, Serialize, Deserialize)]
-pub struct HandleCertificateResponse {
-    pub signed_effects: SignedTransactionEffects,
-}
-
-#[derive(Clone, Debug)]
-pub struct VerifiedHandleCertificateResponse {
-    pub signed_effects: VerifiedSignedTransactionEffects,
-}
-
-#[derive(Clone, Debug, Serialize, Deserialize)]
-pub struct TransactionInfoResponse<
+pub enum TransactionInfoResponse<
     TxnT = SignedTransaction,
     CertT = CertifiedTransaction,
-    EffectsT = SignedTransactionEffects,
+    EfxT = SignedTransactionEffects,
 > {
-    // The signed transaction response to handle_transaction
-    pub signed_transaction: Option<TxnT>,
-    // The certificate in case one is available
-    pub certified_transaction: Option<CertT>,
-    // The effects resulting from a successful execution should
-    // contain ObjectRef created, mutated, deleted and events.
-    pub signed_effects: Option<EffectsT>,
+    Signed(TxnT),
+    // TODO: Eventually support a mode for finalized transactions, where we include raw effects
+    // and the finalized epoch/checkpoint number.
+    // We also shouldn't return the cert in the Executed case, but currently the client is expecting it.
```

Comment:
it would be helpful (if only for debugging purposes) to always have some way to retrieve the original tx from the validator

Reply:
Both Signed and Executed contains some form of this transaction. What case are you referring to?